### PR TITLE
Feature/sci 480 2

### PR DESCRIPTION
**Pull request by timurbazhirov**
Summing up work in #139 and #138 - Rework model hierarchy and add implementation for DFT

### CODE REVIEW

File: schema/models_directory/ml.json [hunk 1]
```diff
@@ -10,14 +10,17 @@
...
-                "ml"
-            ]
-        },
-        "subtype": {
-            "enum": [
-                "re"
-            ]
+            "properties": {
+                "name" : {
+                    "type": "string"
```
Comment (timurbazhirov) [inline schema/models_directory/ml.json:15]: enum: "ml"

File: schema/models_directory/dft/tags.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    "allOf": [
+        {
+            "$ref": "../../model/tags.json"
+        }
+    ],
+    "description": "Tags specific to this model type only.",
+    "items": {
+        "enum": [
+            "excited-states",
+            "scaling-degree:3"
```
Comment (timurbazhirov) [inline schema/models_directory/dft/tags.json:11]: (1) These might not be specific to dft only. Shall we move them to the definitions file (perhaps adding a description there too)?
(2) there must be multiple levels of scaling

File: schema/models_directory/dft/hybrid.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                    {
+                        "$ref": "functional/components/local_functional.json"
+                    },
+                    {
+                        "oneOf": [
+                            {
+                                "$ref": "functional/components/exact_exchange.json"
+                            },
+                            {
+                                "$ref": "functional/components/range_separated_x.json"
```
Comment (timurbazhirov) [inline schema/models_directory/dft/hybrid.json:32]: maybe "range_separated_exchange" to be more clear?

File: schema/models_directory/dft/functional/components/range_separated_x.json [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            "enum": [
+                "range-separated"
+            ]
+        },
+        "type": {
+            "enum": [
+                "exchange"
+            ]
+        },
+        "switchingFrom": {
```
Comment (timurbazhirov) [inline schema/models_directory/dft/functional/components/range_separated_x.json:21]: Maybe "shortRangeFunctional"?

File: schema/models_directory/dft/functional/components/range_separated_x.json [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                },
+                {
+                    "$ref": "functional_names.json#/properties/metaGga"
+                },
+                {
+                    "$ref": "functional_names.json#/properties/exactExchange"
+                }
+            ]
+        },
+        "switchingTo": {
```
Comment (timurbazhirov) [inline schema/models_directory/dft/functional/components/range_separated_x.json:38]: longRangeFunctional?

File: schema/methods_directory/pseudopotential/file/exchange_correlation.json [hunk 6]
```diff
@@ -14,10 +14,10 @@
         "functional": {
             "oneOf": [
                 {
-                    "$ref": "../../../models_directory/dft/definitions/functionals.json#/lda/properties/slug"
+                    "$ref": "../../../models_directory/dft/functional/definitions.json#/lda"
                 },
                 {
-                    "$ref": "../../../models_directory/dft/definitions/functionals.json#/gga/properties/slug"
+                    "$ref": "../../../models_directory/dft/functional/definitions.json#/gga"
```
Comment (timurbazhirov) [inline schema/methods_directory/pseudopotential/file/exchange_correlation.json:20]: Do we only include LDA/GGA here - aren't there pseudopotentials for meta-GGA?

File: example/models_directory/dft/custom_functional.json [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+            "type": "correlation",
+            "fraction": 0.19
+        },
+        {
+            "name": "LYP",
+            "type": "correlation",
+            "fraction": 0.81
+        }
+    ]
+}
```
Comment (timurbazhirov) [inline example/models_directory/dft/custom_functional.json:31]: I suggest moving the lda, gga.json etc. to "functionals_directory" subfolder here, for clarity

File: schema/methods_directory/pseudopotential/file/exchange_correlation.json (hunk 6 above)
Comment (azech-hqs) [reply schema/methods_directory/pseudopotential/file/exchange_correlation.json:20]: I think there are. This was just a minimally invasive change such that the tests would pass.

File: example/models_directory/dft/hybrid.json [hunk 8]
```diff
@@ -0,0 +1,31 @@
+{
+    "approximation": "hybrid",
+    "name": "B3LYP",
```
Comment (azech-hqs) [inline example/models_directory/dft/hybrid.json:3]: These examples are useful for testing. We should clarify how they can be reused with the current implementation of DFT schemas.

**Pull request closed without merging**